### DISCUSSION TITLE
union list[str] or None problem

### DISCUSSION THREAD

**Discussion by fly50789** [fastapi/fastapi#10099] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Union,List
from enum import Enum
from fastapi import FastAPI,Query
from pydantic import BaseModel
from typing_extensions import Annotated

app = FastAPI()

@app.get("/list/items/")
async def read_items_list(q: Union[List[str],None] = Query(default=None)):
    query_items = {"q": q}
    return query_items
```

### Description

In python 3.8.5

If I need to list string or None, I cant add string list in openapi doc.

> q: Union[List[str],List[str]]

![image](https://github.com/tiangolo/fastapi/assets/17740390/60be1240-a08e-4e0c-8218-b73c90536ac7)

> q: Union[List[str],None]

![image](https://github.com/tiangolo/fastapi/assets/17740390/a6549d77-3009-49ed-a3ac-294c79a82d2f)

### Operating System

Windows

### FastAPI Version

0.101.1

### Pydantic Version

2.1.1

### Python Version

Python 3.8.5

**Comment by woldzin** [comment]
Hi, thanks @fly50789 for reporting it!
I also noticed this problem.

The same result for:
- Add Query to Annotated in the q parameter (https://fastapi.tiangolo.com/tutorial/query-params-str-validations/?h=query#add-query-to-annotated-in-the-q-parameter)
- Alternative (old) Query as the default value (https://fastapi.tiangolo.com/tutorial/query-params-str-validations/?h=query#alternative-old-query-as-the-default-value)

screenshot may be misleading:
- Query parameter list / multiple values (https://fastapi.tiangolo.com/tutorial/query-params-str-validations/?h=query#query-parameter-list-multiple-values)

```
Operating System
macOS 13.4.1

FastAPI Version
0.101.1

Pydantic Version
2.1.1

Python Version
Python 3.11.2
```

**Reply by fly50789** [reply]
should I just not use "Union" to aviold this probelm?

BTW another problem in same env.

The limitation cant show correctly.

But it can validate correctly.

```
@app.get("/items/limit/{item_id}")
async def read_items_limit(
    *,
    item_id: int = Path(title="The ID of the item to get", gt=5, le=10),
    q: str,
    size: float = Query(gt=5, lt=10.5),
):
    results = {"item_id": item_id}
    if q:
        results.update({"q": q}) # type: ignore
    return results
```



![image](https://github.com/tiangolo/fastapi/assets/17740390/9e3407bf-b01d-47f1-9d7c-8ebbfb6bdb1b)

![image](https://github.com/tiangolo/fastapi/assets/17740390/baad62fa-e89a-446b-8b0b-36154f71d152)

**Reply by JavierSanchezCastro** [reply, marked as answer]
About the first problem I think it's Swagger side and how they manage and render the data, I suppose it's hard to put in a pretty way all different inputs for the same parameter, also the endpoint is going to work as expected.

The second point seems to be a version mismatch between different tools, there are some open discussions like

https://github.com/tiangolo/fastapi/discussions/9140

https://github.com/OAI/OpenAPI-Specification/issues/1466#issuecomment-493982877

https://stackoverflow.com/questions/76276863/greater-than-gt-and-less-than-lt-parameters-in-fastapi-path-do-not-work-in

https://stackoverflow.com/questions/75862523/fastapis-auto-generated-openapi-file-not-valid-according-to-insomnia

**Reply by fly50789** [reply]
@JavierSanchezCastro 

 Thanks for your support.

Maybe I should aovid this situation until fastapi or swagger make some improvement.
